### PR TITLE
Error Handling

### PR DESCRIPTION
**Pull request by zoefarrell**
This lesson ended up being more complicated to build than I expected, and it ended up being quite a bit of content. I do think that all three parts null checking, validations, and try/catch are important topics so not sure what to cut. Would kind of like to break into two days and then also give them time to implement these things in an API, but not sure if that fits into the schedule well.

### CODE REVIEW

File: module5/lessons/Week1/ErrorHandling.md [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+Today we're going to revisit the MVC Movies Application from Mod 3. Find the repo locally or re-clone it [here](https://github.com/turingschool-examples/MvcMovieStarter). Then check out the branch `read-related-resources`. Start the application up and make sure it runs successfully!
+
+## Ways Applications Can Break!
+
+There are so things that might cause an application to throw an error! 
+
+Let's fill in [this](https://jamboard.google.com/d/1ggm_xNZFJJ1P4qQQ5PngRbSMsobha4cCvZuXHQh0XmY/edit?usp=sharing) Jamboard with some of these ways.
+
+<section class="call-to-action" markdown="1">
+With your partner, take 10 minutes and add to the Jamboard as many ideas as you can for what might cause the application to throw error. Think about both MVC applications and APIs. Feel free to google or experiment with MVC Movies and try to break it!
```
Comment (zoefarrell) [inline module5/lessons/Week1/ErrorHandling.md:35]: I originally was thinking of introducing server side vs client side errors here, but I then this lesson started feeling large so I decided to cut that.

File: module5/lessons/Week1/ErrorHandling.md [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+Or what would happen if someone input a string for a field that can only store an integer? Unless we have some error handling, that would cause an error when we try to write to the database.
+
+The way around these and many other potential problems is to add **Data Validations**. Data validations are checks and rules that are applied to data to ensure its accuracy, integrity, and consistency. The purpose of data validations is to prevent bad data from entering a system, database, or application.
+
+The goal of our data validation is to prevent writing invalid data to the database and instead show users a helpful error message.
+
+![.NET Error](/assets/images/module5/week1/CannotExceed100Chars.png)
+
+
+### Data Validation Step 1: The Model
```
Comment (zoefarrell) [inline module5/lessons/Week1/ErrorHandling.md:200]: This lesson in general feels like it leans more towards sit and get than interactive. I was trying to think about adding more try it out during the lesson, but then I think it would be too long. So I'm thinking we power through these examples and discussions in the morning and then have lots of lab try it out practice time in the afternoon

File: module5/lessons/Week1/ErrorHandling.md (hunk 1 above)
Comment (memcmahon) [reply module5/lessons/Week1/ErrorHandling.md:35]: I agree with that decision - I think you have enough examples of the differences of these below!

File: module5/lessons/Week1/ErrorHandling.md (hunk 2 above)
Comment (memcmahon) [reply module5/lessons/Week1/ErrorHandling.md:200]: I agree - I think with these topics it would be hard to get students to self-research the key topics we want them to get.  I think as long as the lab is pretty exploratory, we should be good here.  Also students will have just come off of intermission, so that might be the best time for them to be able to power through some sit-n-get.

File: module5/lessons/Week1/ErrorHandling.md [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+Let's fill in [this](https://jamboard.google.com/d/1ggm_xNZFJJ1P4qQQ5PngRbSMsobha4cCvZuXHQh0XmY/edit?usp=sharing) Jamboard with some of these ways.
+
+<section class="call-to-action" markdown="1">
+With your partner, take 10 minutes and add to the Jamboard as many ideas as you can for what might cause the application to throw error. Think about both MVC applications and APIs. Feel free to google or experiment with MVC Movies and try to break it!
+
+The 400 error codes here are good place to start for API errors! [HTTP Cats](https://http.cat/)
+</section>
+
+<aside class="instructor-notes" markdown="1">
+    <p><strong>Instructor Note</strong><br>
```
Comment (memcmahon) [inline module5/lessons/Week1/ErrorHandling.md]: Love this list of examples - I wonder if we should prompt instructors to start the jamboard with one or two of these to get students heads in the right space?

File: module5/lessons/Week1/ErrorHandling.md [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+</section>
+
+Let's add some checks to make sure those values are not null.
+
+```c#
+[Route("Movies/{id:int}")]
+public IActionResult Show(int id)
+{
+    
+    if (id == null)
```
Comment (memcmahon) [inline module5/lessons/Week1/ErrorHandling.md:160]: I think the way this action is set up, the id could never come in as null.  I played around with it, and if I send in a non-integer, I get a 404; if I send nothing, it goes to the index action.  Also, I think the default value of an int is `0` so if someone got through to this action with any non-integer, I think it would be 0 and not null?

I could also be testing this wrong, so if you got a null value here, you can ignore me :)

File: module5/lessons/Week1/ErrorHandling.md [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+    <p><strong>Instructor Note</strong><br>
+    This code will hook up to the model and display the current value if there is one.
+    <label asp-for="Title">Title:</label>
+    <input asp-for="Title" class="form-control" />
+    </p>
+</aside>
+
+It's out of scope of this lesson, but if you want to display what the user previous input I recommend looking into 
+.NET tag helpers, specifically [asp-for](https://learn.microsoft.com/en-us/aspnet/core/mvc/views/working-with-forms?view=aspnetcore-5.0#the-input-tag-helper).
+
```
Comment (memcmahon) [inline module5/lessons/Week1/ErrorHandling.md:292]: I think we should call out the difference in the server-side validations, and client side validations here.  Something like 'using the asp-validation-for send instructions to the form to prevent 'bad' data from being submitted`.  We could re-inforce the client-side nature by inspecting this form to see the raw html that is controlling these validations.

File: module5/lessons/Week1/ErrorHandling.md [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+{
+    Console.WriteLine($"Error: The file '{filePath}' was not found.");
+}
+```
+
+Because we are "catching" the error that occurred in the try block, the program will not crash it will just log the error!
+
+#### Don't Swallow Errors
+
+Take a look at this second example of a try/catch block. This is an example of what in the software engineering industry gets referred to as "Swallowing an error"
```
Comment (memcmahon) [inline module5/lessons/Week1/ErrorHandling.md:328]: Ooh - I was not familiar with swallowing errors; cool!!

File: module5/preparation/Week1/ErrorHandling.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+To prepare for the lesson on Error Handling, read [Writing Helpful Error Messages](https://developers.google.com/tech-writing/error-messages) from the Google Technical Writing Course to help you start thinking about what goes into good and bad error handling.
+
+Then read the following Error Handling Rules:
+* [Don't Fail Silently](https://developers.google.com/tech-writing/error-messages/error-handling#dont_fail_silently)
+* [Avoid Swallowing the Root Cause](https://developers.google.com/tech-writing/error-messages/error-handling#avoid_swallowing_the_root_cause)
+* [Log The Error Codes](https://developers.google.com/tech-writing/error-messages/error-handling#log_the_error_codes)
+* [Raise Errors Immediately](https://developers.google.com/tech-writing/error-messages/error-handling#raise_errors_immediately)
+
+
+✅ Deliverable: ???????????? TBD
```
Comment (memcmahon) [inline module5/preparation/Week1/ErrorHandling.md]: I wouldn't stress too much about a deliverable - as we get further in the program, this could fall under 'release of responsibility' 🙃

**Comment by memcmahon** [comment]
I agree with your assessment of the size and necessity of this lesson - I moved around the calendar just a bit to give us a full morning on this lesson... I think we could also move things around to make it two days, but if we think we can get away with one day, that would be my preference.

File: module5/lessons/Week1/ErrorHandling.md (hunk 3 above)
Comment (zoefarrell) [reply module5/lessons/Week1/ErrorHandling.md]: Great minds think alike! I already added two stickies to the Jamboard of what we're looking for :) But would probably be helpful for future cohorts if I include an instructor note to do this.

File: module5/preparation/Week1/ErrorHandling.md (hunk 7 above)
Comment (zoefarrell) [reply module5/preparation/Week1/ErrorHandling.md]: I like that perspective

**Comment by zoefarrell** [comment]
> I agree with your assessment of the size and necessity of this lesson - I moved around the calendar just a bit to give us a full morning on this lesson... I think we could also move things around to make it two days, but if we think we can get away with one day, that would be my preference.

Yeah, I think it'll be ok as one day, but a big day. Let's try it as one and then if it's really too much next time we can break it into two.

File: module5/lessons/Week1/ErrorHandling.md (hunk 4 above)
Comment (zoefarrell) [reply module5/lessons/Week1/ErrorHandling.md:160]: This is an interesting question. I took this from the code the Entity Framework .NET docs use which looks like this. 

```
// GET: Students/Details/5
        public async Task<IActionResult> Details(int? id)
        {
            if (id == null)
            {
                return NotFound();
            }

            var student = await _context.Students
                .FirstOrDefaultAsync(m => m.ID == id);
            if (student == null)
            {
                return NotFound();
            }

            return View(student);
        }
```

But after looking closer they always have their parameters as optional. I'm not entirely sure why because they never do anything if the id isn't provided besides throwing a 404. Maybe better to get to this route and throw a 404 than to not match a route? And maybe in the future some nicer error handling would get added here?

I added a bit extra here to show only one null check needed if the parameter is required, and two if it is optional.

File: module5/labs/Week1/ErrorHandling.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+title: Error Handling Lab
+---
+
+In today's lab, we are going to continue adding Error Handling to our MVC Movies Application.
+
+### Step 1: Adding Null Checking
+
+First check out a new branch. Then add null checking throughout the rest of the application.
+
+Once you think your application is thoroughly null-checked, create a PR and have a code review session with a peer.
```
Comment (memcmahon) [inline module5/labs/Week1/ErrorHandling.md:12]: Yes!! Love the PR process included here

File: module5/labs/Week1/ErrorHandling.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+First check out a new branch. Then spend five minutes researching "Sad path testing". Send a description of sad path testing in your own words to all instructors.
+
+Then add two "Sad path" tests to your application.
+
+Once you have added your tests, create a PR and have a code review session with a peer.
+
+
+<aside class="instructor-notes" markdown="1">
+    <p><strong>Instructor Note</strong><br>
+    If students finish early and are looking for more to do, you could point them towards researching/implementing a global exception handler or a custom error page.</p>
```
Comment (memcmahon) [inline module5/labs/Week1/ErrorHandling.md:31]: I think it will be interesting to see how this timing plays out.  I think there is enough space in the lab that it won't be a catastrophe if the lesson goes over into the afternoon.

**Review by memcmahon — APPROVED** [review]
Excited to see how this goes!  This is a new space for Turing and I'm stoked we are covering it in Launch

**Comment by memcmahon** [comment]
Let's add a link to this lesson from the index, and then I think we are good to go!
